File: python-sdk/src/sayna_client/credentials.py
"""Resolve GoogleAuth credentials from strings or file paths."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any, overload

from sayna_client.errors import SaynaValidationError
from sayna_client.types import (
    GoogleAuth,
    ProviderAuth,
    STTConfig,
    TTSConfig,
)


def resolve_google_credentials(
    credentials: str | dict[str, Any],
) -> dict[str, Any]:
    """Resolve Google credentials from a string to a parsed dict.

    If the value is already a dict it is returned as-is.
    If it is a string, the function first tries ``json.loads``.
    When parsing fails, the string is treated as a file path
    and the file is read and parsed.

    Raises:
        SaynaValidationError: When the string is neither valid JSON
            nor a readable JSON file.
    """
    if isinstance(credentials, dict):
        return credentials

    # Try parsing as inline JSON string
    try:
        parsed = json.loads(credentials)
        if isinstance(parsed, dict):
            return parsed
        raise SaynaValidationError(
            f"GoogleAuth credentials JSON string must parse to an object, got {type(parsed).__name__}"
        )
    except json.JSONDecodeError:
        pass  # Not valid JSON — fall through to file path handling

    # Treat as file path
    path = Path(credentials)
    try:
        content = path.read_text(encoding="utf-8")
    except OSError as exc:
        msg = (
            "GoogleAuth credentials string is not valid JSON "
            f"and could not be read as a file: {credentials}"
        )
        raise SaynaValidationError(msg) from exc

    try:
        parsed = json.loads(content)
    except json.JSONDecodeError as exc:
        raise SaynaValidationError(
            f"GoogleAuth credentials file does not contain valid JSON: {credentials}"
        ) from exc

    if not isinstance(parsed, dict):
        raise SaynaValidationError(
            f"GoogleAuth credentials file must contain a JSON object: {credentials}"
        )

    return parsed


def resolve_provider_auth(auth: ProviderAuth) -> ProviderAuth:
    """Resolve provider auth credentials.

    For :class:`GoogleAuth` with string credentials the value is resolved
    via :func:`resolve_google_credentials`. All other auth types pass through
    unchanged.
    """
    if not isinstance(auth, GoogleAuth) or not isinstance(auth.credentials, str):
        return auth
    return GoogleAuth(credentials=resolve_google_credentials(auth.credentials))


@overload
def resolve_config_auth(config: STTConfig) -> STTConfig: ...


@overload
def resolve_config_auth(config: TTSConfig) -> TTSConfig: ...


@overload
def resolve_config_auth(config: None) -> None: ...


def resolve_config_auth(
    config: STTConfig | TTSConfig | None,
) -> STTConfig | TTSConfig | None:
    """Resolve provider auth inside an STT or TTS config object.

    Returns the original config when no resolution is needed,
    or a copy with the resolved auth otherwise.
    Accepts ``None`` for convenience and returns it unchanged.
    """
    if config is None or config.auth is None:
        return config
    resolved = resolve_provider_auth(config.auth)
    if resolved is config.auth:
        return config
    return config.model_copy(update={"auth": resolved})

File: node-sdk/src/credentials.ts
import { readFile } from "node:fs/promises";
import type { GoogleAuth, ProviderAuth } from "./types";
import { SaynaValidationError } from "./errors";

function isGoogleAuth(auth: ProviderAuth): auth is GoogleAuth {
  return "credentials" in auth;
}

/**
 * Resolves Google credentials from a string to a parsed object.
 *
 * If the value is already an object it is returned as-is.
 * If it is a string, the function first tries `JSON.parse`.
 * When parsing fails, the string is treated as a file path
 * and the file is read and parsed.
 *
 * @throws {SaynaValidationError} When the string is neither valid JSON nor a readable JSON file.
 */
export async function resolveGoogleCredentials(
  credentials: string | Record<string, unknown>
): Promise<Record<string, unknown>> {
  if (typeof credentials !== "string") {
    return credentials;
  }

  // Try parsing as inline JSON string
  try {
    const parsed: unknown = JSON.parse(credentials);
    if (
      typeof parsed === "object" &&
      parsed !== null &&
      !Array.isArray(parsed)
    ) {
      return parsed as Record<string, unknown>;
    }
    throw new SaynaValidationError(
      `GoogleAuth credentials JSON string must parse to an object, got ${Array.isArray(parsed) ? "array" : typeof parsed}`
    );
  } catch (error) {
    if (error instanceof SaynaValidationError) {
      throw error;
    }
    // Not valid JSON — fall through to file path handling
  }

  // Treat as file path
  let content: string;
  try {
    content = await readFile(credentials, "utf-8");
  } catch (_error) {
    throw new SaynaValidationError(
      `GoogleAuth credentials string is not valid JSON and could not be read as a file: ${credentials}`
    );
  }

  let parsed: unknown;
  try {
    parsed = JSON.parse(content);
  } catch {
    throw new SaynaValidationError(
      `GoogleAuth credentials file does not contain valid JSON: ${credentials}`
    );
  }

  if (typeof parsed !== "object" || parsed === null || Array.isArray(parsed)) {
    throw new SaynaValidationError(
      `GoogleAuth credentials file must contain a JSON object: ${credentials}`
    );
  }

  return parsed as Record<string, unknown>;
}

/**
 * Resolves provider auth credentials.
 *
 * For {@link GoogleAuth} with string credentials the value is resolved
 * via {@link resolveGoogleCredentials}. All other auth types pass through unchanged.
 */
export async function resolveProviderAuth(
  auth: ProviderAuth
): Promise<ProviderAuth> {
  if (!isGoogleAuth(auth) || typeof auth.credentials !== "string") {
    return auth;
  }
  return {
    credentials: await resolveGoogleCredentials(auth.credentials),
  };
}

/**
 * Resolves provider auth inside an STT or TTS config object.
 *
 * Returns the original config when no resolution is needed,
 * or a shallow copy with the resolved auth otherwise.
 * Accepts `undefined` for convenience and returns it unchanged.
 */
export async function resolveConfigAuth<
  T extends { auth?: ProviderAuth },
>(config: T | undefined): Promise<T | undefined> {
  if (!config?.auth) {
    return config;
  }
  const resolved = await resolveProviderAuth(config.auth);
  if (resolved === config.auth) {
    return config;
  }
  return { ...config, auth: resolved };
}
